### PR TITLE
Scala 3: add support for extension methods

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, Scalafmt didn't support formatting extension methods.
This commit fixes that issue for code using braces. There's still no
support for indentation syntax (which is still being designed and is
likely to change soon).

### CODE REVIEW

File: scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala [hunk 1]
```diff
@@ -54,6 +54,9 @@ class FormatTests extends FunSuite with CanRunTests with FormatAssertions {
         e.partialOutput
       case x => x.get
     }
+    if (!onlyManual) {
+      assertEquals(obtained, t.expected, "Obtained formatting output does not match the expected formatting output" )
```
Comment (olafurpg) [inline scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala:58]: `assertEquals` produces copy-pasteable output instead of `.stripMargin` formatted multiline strings from `assertNoDiff`

File: project/Dependencies.scala [hunk 2]
```diff
@@ -6,7 +6,7 @@ import org.portablescala.sbtplatformdeps.PlatformDepsPlugin.autoImport._
 
 object Dependencies {
   val metaconfigV = "0.9.10"
-  val scalametaV  = "4.4.2"
+  val scalametaV  = "4.4.3"
```
Comment (olafurpg) [inline project/Dependencies.scala:9]: This release includes an important fix for positions

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 3]
```diff
@@ -464,6 +464,14 @@ class Router(formatOps: FormatOps) {
           Split(NewlineT(isDouble = tok.hasBlankLine), 0)
         )
 
+      case tok @ FormatToken(_: T.RightParen, _: T.KwDef, _)
+          if leftOwner.is[Defn.ExtensionGroup] =>
+        // Force a newline with the body of a Defn.ExtensionGroup if the
+        // method doesn't fit on a single line.
+        Seq(
+          Split(Space, 0).withSingleLine(leftOwner.tokens.last),
+          Split(Newline, 1).withIndent(2, leftOwner.tokens.last, After)
```
Comment (olafurpg) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: The 2 space indentation is hardcoded in several places so I hardcoded the `2` literal here as well. We may need to make this indentation configuration down the road with significant indentation syntax.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenClasses.scala [hunk 4]
```diff
@@ -4,6 +4,16 @@ import scala.meta.internal.classifiers.classifier
 import scala.meta.tokens.Token
 import scala.meta.tokens.Token._
 
+@classifier
+trait ExtensionKeyword
+object ExtensionKeyword {
+  def unapply(token: Token): Boolean = token match {
+    case _: KwExtension => true
+    case Ident("extension") => true
```
Comment (tgodzik) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenClasses.scala:12]: So `extension` is still a soft keyword, but it was confirmed that is should be changed as it's impossible distinguish. I reopened the issue in Dotty, since it was closed and plainly ignored https://github.com/lampepfl/dotty/issues/10076

In case it becomes a keyword this case should not be needed and currently scalameta always parses extension as a keyword.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 5]
```diff
@@ -137,6 +151,8 @@ object TreeOps {
         case t: Defn.Macro => addDefn[KwDef](t.mods, t)
         case t: Decl.Def => addDefn[KwDef](t.mods, t)
         case t: Defn.Enum => addDefn[KwEnum](t.mods, t)
+        case t: Defn.ExtensionGroup =>
+          addDefnTokens(Nil, t, "extension", t => ExtensionKeyword.unapply(t))
```
Comment (tgodzik) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala:155]: Why not just `addDefn[KwExtension](Nil, t)` ?

**Review by tgodzik — CHANGES_REQUESTED** [review]
I think we should currently assume that extension is a keyword. We can change it back in the future, but that's how it will be currently tokenized for the Dotty dialect.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 6]
```diff
@@ -464,6 +464,14 @@ class Router(formatOps: FormatOps) {
           Split(NewlineT(isDouble = tok.hasBlankLine), 0)
         )
 
+      case tok @ FormatToken(_: T.RightParen, _: T.KwDef, _)
```
[the remark this answers is not in the document]
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:467]: one other thing: what about
```
extension [A](...) /* new feature */ // or this
  def ...
```

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 3 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: i think significant is only important for parsing. for formatting, you can fix it to 2, no?

**Review by kitbellew — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 7]
```diff
@@ -464,6 +464,14 @@ class Router(formatOps: FormatOps) {
...
+          if leftOwner.is[Defn.ExtensionGroup] =>
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:468]: It will not be parsed as an extension if the dialect does not support it.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenClasses.scala (hunk 4 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenClasses.scala:12]: I cannot promise that it will be enough for the future, since there is still changes being made, but I am 90 % sure. There is a number of cases that extension as a soft keyword will not just work like:
```
extension(a : Int){
  def b = a
}
```

This can be an extension or invocation of `def extension(a : Int)( c : => Unit)`

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenClasses.scala:12]: if we end up keeping this logic, i think that
- this commit needs to be rebased on the `master`
- this trait moved inside @MichalBednarz's new `SoftKeywordClasses` type
- `true` on this line replaced with `dialect.allowExtensionMethods`.

Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TokenClasses.scala:12]: So actually, it will be a soft keyword :/ I was told otherwise and now again they changed their minds. Sorry about that, I will fix it in the parser.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/DanglingParentheses.scala [hunk 8]
```diff
@@ -36,10 +36,11 @@ object DanglingParentheses {
     case object `class` extends Exclude
     case object `trait` extends Exclude
     case object `enum` extends Exclude
+    case object `extension` extends Exclude
     case object `def` extends Exclude
 
     implicit val reader: ConfCodec[Exclude] =
-      ReaderUtil.oneOf[Exclude](`class`, `trait`, `enum`, `def`)
+      ReaderUtil.oneOf[Exclude](`class`, `trait`, `enum`, `extension`, `def`)
```
Comment (MichalBednarz) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/DanglingParentheses.scala:43]: A corresponding part in the documentation can be updated. In `danglingParentheses.exclude` it's stated that `currently, class, trait and def are supported`. I haven't included there the enum by myself. You can add both extension and enum if you want.

Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/DanglingParentheses.scala:43]: Added it!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 9]
```diff
@@ -464,6 +464,14 @@ class Router(formatOps: FormatOps) {
...
+        // Force a newline with the body of a Defn.ExtensionGroup if the
+        // method doesn't fit on a single line.
+        Seq(
+          Split(Space, 0).withSingleLine(leftOwner.tokens.last),
+          Split(Newline, 1).withIndent(2, leftOwner.tokens.last, After)
+        )
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:480]: I added it. I also updated indent `expire` to take into account comments

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala (hunk 5 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Actually, this was needed, but I changed it to reuse SoftKEywords class.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 10]
```diff
@@ -123,12 +128,21 @@ object TreeOps {
         case Some(x) => addTree(x, tree)
         case _ =>
           // No non-annotation modifier exists, fallback to keyword like `object`
-          tree.tokens.find(classTag[T].runtimeClass.isInstance) match {
+          tree.tokens.find(isMatch) match {
             case Some(x) => addTok(x, tree)
-            case None => throw Error.CantFindDefnToken[T](tree)
+            case None =>
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Fixed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 11]
```diff
@@ -114,7 +114,12 @@ object TreeOps {
       t.tokens.find(!_.is[Trivia]).foreach(addTok(_, tree))
     def addAll(trees: Seq[Tree]) = trees.foreach(x => addTree(x, x))
 
-    def addDefn[T: ClassTag](mods: Seq[Mod], tree: Tree): Unit = {
+    def addDefnTokens[T](
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Removed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 12]
```diff
@@ -123,12 +128,21 @@ object TreeOps {
...
+              throw Error.CantFindDefnToken(what, tree)
           }
       }
     }
+    def addDefn[T: ClassTag](mods: Seq[Mod], tree: Tree): Unit = {
+      addDefnTokens[T](
+        mods,
+        tree,
+        classTag[T].runtimeClass.getSimpleName(),
+        t => classTag[T].runtimeClass.isInstance(t)
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Changed it a bit taking into account your comment, thanks!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: Good catch, I added support for the edge case with comments and added some tests, let me know what you think!

**Comment by tgodzik** [comment]
@olafurpg @kitbellew I updated the PR after fixing `extension` to be a soft keyword again in scalameta. I also added some tests to make sure it all works, let me know what you think!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 13]
```diff
@@ -561,6 +566,8 @@ class Router(formatOps: FormatOps) {
               if style.spaces.afterSymbolicDefs &&
                 isSymbolicName(name.value) && name.parent.exists(isDefDef) =>
             Space
+          case ext: Defn.ExtensionGroup if tok.left.is[soft.KwExtension] =>
```
Comment (MichalBednarz) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:569]: nitpick: `_: Defn.ExtensionGroup`

File: scalafmt-tests/src/test/resources/dotty/Extension.stat [hunk 14]
```diff
@@ -94,3 +94,50 @@ extension [AAaaaaaaaaaa](
...
+===
+extension [A](cat: Cat){ /* new feature */ // or this
+  def a = 1
+}
+>>>
+extension [A](
+    cat: Cat
+) { /* new feature */ // or this
+  def a = 1
+}
```
Comment (MichalBednarz) [inline scalafmt-tests/src/test/resources/dotty/Extension.stat:143]: add newline

**Review by MichalBednarz — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 15]
```diff
@@ -466,6 +466,19 @@ class Router(formatOps: FormatOps) {
           Split(NewlineT(isDouble = tok.hasBlankLine), 0)
         )
 
+      case tok @ FormatToken(_: T.RightParen, _: T.KwDef | _: T.Comment, _)
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:469]: what happens if right is comment but after that we don't get a KwDef?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 16]
```diff
@@ -466,6 +466,19 @@ class Router(formatOps: FormatOps) {
...
+          if leftOwner.is[Defn.ExtensionGroup] =>
+        val expireToken = tok.right match {
+          case _: T.KwDef => leftOwner.tokens.last
+          case _ => nextNonComment(next(tok)).right
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: why right? looks inconsistent with the KwDef handling.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 17]
```diff
@@ -527,7 +540,7 @@ class Router(formatOps: FormatOps) {
         )
       // Opening [ with no leading space.
       // Opening ( with no leading space.
-      case FormatToken(
+      case tok @ FormatToken(
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:543]: i wouldn't add tok. you can always refer to the actual variable, formatToken.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 13 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:569]: yep

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala [hunk 18]
```diff
@@ -137,6 +155,8 @@ object TreeOps {
         case t: Defn.Macro => addDefn[KwDef](t.mods, t)
         case t: Decl.Def => addDefn[KwDef](t.mods, t)
         case t: Defn.Enum => addDefn[KwEnum](t.mods, t)
+        case t: Defn.ExtensionGroup =>
+          addDefnTokens(Nil, t, "extension", t => soft.KwExtension.unapply(t))
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala:159]: i may have asked. why do we need a lambda here? can't we just use KwExtension.unapply directly?

File: scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala [hunk 19]
```diff
@@ -54,6 +54,13 @@ class FormatTests extends FunSuite with CanRunTests with FormatAssertions {
         e.partialOutput
       case x => x.get
     }
+    if (!onlyManual) {
```
Comment (kitbellew) [inline scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala:57]: why did you reorder? i think the original idea was to put higher preference on idempotence violations, although I can't remember why. generally, if you believe this way is better, I'd strongly recommend doing that as part of a separate pull request, so the intent is clearly recorded in history.

File: scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala [hunk 20]
```diff
@@ -74,10 +81,7 @@ class FormatTests extends FunSuite with CanRunTests with FormatAssertions {
       )
       .get
     debug2.printTest()
-    assertNoDiff(formattedAgain, obtained, "Idempotency violated")
-    if (!onlyManual) {
-      assertNoDiff(obtained, t.expected)
-    }
+    assertEquals(formattedAgain, obtained, "Idempotency violated")
```
Comment (kitbellew) [inline scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala:84]: looks like we're also changing one type of assert to another. can we do that in a separate pr? i think it deserves its own commit, with explicit explanation.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala [hunk 21]
```diff
@@ -88,6 +88,10 @@ object ScalafmtRunner {
       .withAllowTrailingCommas(true)
     val default = scala213
     val Scala3 = Dotty
+      // Disable significant indentation because the exact syntax is still under
+      // development and is likely to change soon. See discussions at
+      // https://contributors.scala-lang.org/t/feedback-sought-optional-braces/4702
+      .withAllowSignificantIndentation(false)
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:95]: not a big fan of latent changes.

**Review by kitbellew — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala (hunk 18 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala:159]: The problem is that `extension` is a soft keyword so a tokenizer treats it as an Ident and `ident` will not match `extension`

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala (hunk 21 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:95]: Unfortunately, I think until it all is stable we shouldn't enable it in scalafmt.

File: scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala (hunk 19 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala:57]: Not sure what @olafurpg meant here, but I guess I can remove the changes.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala (hunk 21 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:95]: I created a separate PR for some of the unrelated changes in this PR: https://github.com/scalameta/scalafmt/pull/2312

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 15 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:469]: That cannot happen, the code will not parse then.  Nothing significant can be added between `)` and `def`

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 16 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: `right` will be the `def` token. It doesn't work with `left`, which will be still a comment.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala (hunk 21 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtRunner.scala:95]: I will rebase once the change is made.

File: scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala (hunk 20 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala]: Moved the change to https://github.com/scalameta/scalafmt/pull/2312

File: scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala (hunk 19 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/scala/org/scalafmt/FormatTests.scala]: Moved to https://github.com/scalameta/scalafmt/pull/2312

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 17 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: Fixed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 22]
```diff
@@ -466,6 +466,19 @@ class Router(formatOps: FormatOps) {
           Split(NewlineT(isDouble = tok.hasBlankLine), 0)
         )
 
+      case tok @ FormatToken(_: T.RightParen, _: T.KwDef | _: T.Comment, _)
+          if leftOwner.is[Defn.ExtensionGroup] =>
+        val expireToken = tok.right match{
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: Fixed

File: scalafmt-tests/src/test/resources/dotty/Extension.stat (hunk 14 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/resources/dotty/Extension.stat]: Fixed

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 13 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: Fixed

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 15 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:469]: in that case, does it still make sense to match `KwDef` or `Comment` here? shouldn't `_` be equivalent, given the `leftOwner` check below?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 16 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:473]: so for comment, you will return `KwDef`. but in case of `KwDef`, you are not returning the `KwDef` token, you are returning the last token of the entire `ExtensionGroup`.

wouldn't `leftOwner.tokens.last` be sufficient for both cases? or `lastToken(leftOwner)` (which stops before trailing comments/whitespace)? i am not clear why there's asymmetry here.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala (hunk 18 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala:159]: ok, i am not sure i understood. what i was asking was whether this code would work equally:
```
  addDefnTokens(Nil, t, "extension", soft.KwExtension.unapply)
```
that is, instead of `t => soft.KwExtension.unapply(t)`, can we use the simpler `soft.KwExtension.unapply`. and if we can't, can we at least use a different variable (because the case also uses `t`, so it might be a bit confusing)?

**Review by kitbellew — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 15 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:469]: Ach, actually you can have a using clause `(using A)` and an opening bracket `{`  We want to format it differently in that case.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 16 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: I was overthinking it, `leftOwner.tokens.last` works for both cases, thanks!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 15 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:469]: If we leave it as just `_` it will format it a bit weirdly and also break the parser, since there is an issue when the formatting is a bit off. I will fix it separately, but the way it's currently formatted looks good to me.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala (hunk 18 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/util/TreeOps.scala]: Ach, sorry! I totally misread it - or in reality I somehow never really read it :O  Changed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala (hunk 15 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala:469]: > Ach, actually you can have a using clause `(using A)` and an opening bracket `{` We want to format it differently in that case.

we can merge this as-is, but it would definitely be useful to have a follow-up pr to handle this:
```
extension [A, B](
  a: A
) /* foo */ (
  using b: B /* bar */
) /* baz */ {
  def a = 1
}
```

**Review by kitbellew — APPROVED** [review]
